<h2> Комментарий ревьюера <a class="tocSkip"></h2>

Владимир, привет!

Меня зовут Олег Михеев, и я буду проверять твой проект. Предлагаю общаться на "ты", но если тебе будет некомфортно, обязательно сообщи, и мы перейдем на "вы". Моя главная цель - не указать на твои ошибки, а поделиться опытом и подготовить тебя к работе аналитиком. Увидев у тебя ошибку, я постараюсь указать на её наличие и дам возможность найти и самостоятельно исправить её.

В проверке проекта я буду использовать разные цвета для обозначения комментариев. Например:

<br/>

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Так выделены удачные решения, на которые можно опираться в будущем.
</div>

<br/>

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Когда решение на отдельном шаге станет лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

 А так выделены решения, которые необходимо доработать, чтобы я мог принять проект.
</div>
    
  
Если вносишь изменения в проект, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь их заметным цветом. Например, вот так:
       
<div class="alert alert-block alert-info">
    
<h3> Комментарий студента  <a class="tocSkip"></h3>
    
Можешь кликнуть сюда два раза и скопировать код комментария. Также если что-то непонятно или остались вопросы по ревью, обязательно пиши.
    
</div>
    
Пожалуйста, не удаляй и не перемещай мои комментарии, чтобы не произошло путаницы, и я мог быстрее проверить проект.
<br/>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью**

Хорошая работа. Ты умело пользуешься кодом для решения поставленных задач.
    
А твои выводы понятны и отражают результаты расчетов.

**Что нужно сделать, чтобы я мог принять проект**:

* Написать подробный финальный вывод, учитывающий все группы клиентов с процентными показателями и информацию о предобработке

**Успехов! Жду проект на повторное ревью!**

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью v2**

Спасибо, что учёл замечание!

Очень качественно выполненный проект! Мне было приятно с тобой работать 🤝

Видно, что ты последовательно и вдумчиво подходишь к анализу.
    
А проект написан аккуратно, поэтому его легко воспринимать.

Продолжай в том же духе и станешь крутым специалистом.

**Успехов в дальнейшем обучении!**

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

 <div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Чтобы не приходилось несколько раз выполнять одну и ту же операцию, можно автоматизировать процесс вывода сводных таблиц с помощью пользовательской функции. Например, такой:

 ```py
def p_table_create(category):
  table = data.groupby(category)['debt'].agg(['count', 'sum', 'mean'])\
    .sort_values(by = 'mean', ascending = False)  
  table['mean'] = table['mean'].apply(lambda x: "{0:.2%}".format(x))
  return table
```
    
</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

#### Для этого сопоставим людей с детьми, с людьми у которых детей нет, и сравним кто возвращает кредиты быстрее.

In [57]:
pivot_table_c = data.pivot_table(index='children', columns='debt', values='total_income', aggfunc='count') # создаём сводную таблицу

pivot_table_c.columns = ['in_time', 'not_in_time'] # переименовываем колонки для удобства

pivot_table_c = pivot_table_c.fillna(0) #заполним NaN нулями(если есть NaN)

In [55]:
total_people_loans = pivot_table_c['in_time'] + pivot_table_c['not_in_time'] # считаем общее число кредитов

pivot_table_c.insert(0, 'total_loans', total_people_debts) # добавим колонку с общим числом займов

pivot_table_c['percent_in_time'] = (pivot_table_c['in_time'] / total_people_debts) * 100 # считаем процент тех людeй, кто возвращает кредит в срок и добавляем столбец в таблицу 

pivot_table_c # выводим результат на экран

total_loans  in_time  not_in_time  percent_in_time
children                                                    
0             14091.0  13028.0       1063.0        92.456178
1              4808.0   4364.0        444.0        90.765391
2              2052.0   1858.0        194.0        90.545809
3               330.0    303.0         27.0        91.818182
4                41.0     37.0          4.0        90.243902
5                 9.0      9.0          0.0       100.000000

### Вычислим средний процент возвращенных во время кредитов, отбросив людей без детей и с 5-ю детьми, так как у людей с 5-ю детьми всего 9 кредитов и у нас отсутствовали данные о просрочках, что может также означать что просрочки имеются, но по каким-то причинам отсутствуют, что может повлиять на все вычисления целиком.


In [24]:
filtered_pivot_table_с = pivot_table_c.loc[1:4] # отфильтруем значения людей с детьми, и людей с 5-ю детьми

with_children_avg = filtered_pivot_table_с['percent_in_time'].mean() # средний процент возвращенных в срок кредитов, людьми с детьми

print(f"Люди с детьми, в среднем, возвращают в срок {round(with_children_avg, 2)}% кредитов") # выводим результат на экран

Люди с детьми, в среднем, возвращают в срок 90.84% кредитов


In [25]:
relation = pivot_table_c.loc[0, 'percent_in_time'] - with_children_avg # разница между людьми без детей и с детьми, возвращающими кредит в срок.

print(f"Люди с детьми возвращают кредиты в срок на {round(relation, 1)}% реже, чем без детей") # выводим результат на экран

Люди с детьми возвращают кредиты в срок на 1.6% реже, чем без детей


**Вывод:** Исходя из получившейся сводной таблицы, а также исходя из подсчетов, видно что в среднем, люди с детьми выплачивают кредиты в срок на **1.6%** **реже** чем без детей, что отвечает на вопрос, поставленный в задаче(зависимость есть). Рекомендуется провести дополнительное исследование для определения значимости данной информации при формировании скоринговой системы.


 <div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

У тебя получилась отличная сводная таблица со всеми необходимыми данными.

 Вывод верный. Здорово, что обращаешь внимание на размеры категорий клиентов. Так можно быть уверенным, что выводы объективны

 Можно также сделать расчёты с помощью нескольких функций по одному или нескольким столбцам.
    
Примеры:
    
* data.groupby('column1')['column2'].agg(['count', 'mean', 'sum'])
    
* data.groupby('column1').agg({'column1' : 'count', 'column2' : 'mean'})
    
* data.pivot_table(index = 'column1' , values = 'column2', aggfunc=['count', 'mean', 'sum'])
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

#### Для этого создадим сводную таблицу, и сопоставим разные семейные статусы с "контрольной группой", 

In [46]:
pivot_table_family = data.pivot_table(index='family_status', columns='debt', values='total_income', aggfunc='count') # создаём сводную таблицу

pivot_table_family.columns = ['in_time', 'not_in_time'] # переименовываем колонки для удобства
pivot_table_family = pivot_table_family.sort_values(by='in_time', ascending=False) # сортируем значения по убыванию для удобства и сохраняем получившуюся таблицу

total_loans_family = pivot_table_family['in_time'] + pivot_table_family['not_in_time'] # сохраняем общее число займов.
pivot_table_family.insert(0, 'total_loans', total_loans) # вставим колонку с общим кол-вом займов вначало таблицы 

pivot_table_family # выводим таблицу на экран

total_loans  in_time  not_in_time
family_status                                           
женат / замужем              12261    11334          927
гражданский брак              4134     3749          385
Не женат / не замужем         2796     2523          273
в разводе                     1189     1105           84
вдовец / вдова                 951      888           63

In [48]:
pivot_table_family['percent_in_time'] = pivot_table_family['in_time'] / total_loans_family * 100 # считаем процент возвращенных в срок кредитов
pivot_table_family = pivot_table_family.sort_values(by='percent_in_time', ascending=True) # сортируем значения по проценту возврата кредита в срок, по возрастанию

pivot_table_family # выводим результат на экран D

total_loans  in_time  not_in_time  percent_in_time
family_status                                                            
Не женат / не замужем         2796     2523          273        90.236052
гражданский брак              4134     3749          385        90.686986
женат / замужем              12261    11334          927        92.439442
в разводе                     1189     1105           84        92.935240
вдовец / вдова                 951      888           63        93.375394

#### Здесь предлагаем сравнить разные группы между собой, для более наглядного представления.

In [49]:
# посчитаем средний процент возвращенных кредитов в группах. Сравним людей, которые никогда не вступали в брак, со всеми остальными. 
f
iltered_family_table = pivot_table_family[1:5] # отфильтруем таблицу, включив в неё людей кто состоит в браке или когда-то в нём состоял.

avg_family_percent = filtered_family_table['percent_in_time'].mean() # вычисляем среднее значение возвращенных в срок кредитов среди людей состоящих или бывших в браке.

print(f"Люди состоящие или бывшие в браке, в среднем, возвращают кредиты в срок {round(avg_family_percent, 2)}% случаев") # выводит результат на экран

Люди состоящие или бывшие в браке, в среднем, возвращают кредиты в срок 92.36% случаев


#### Как мы можем видеть по расчетам, люди, состоящие или бывшие в браке, в среднем, возвращают кредиты в срок 92.36% случаев

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Все выводы и мысли по проекту принято выносить в ячейки маркдаун. Комментарии в ячейках с кодом предназначены для описания работы именно кода.

</div>

In [50]:
family_status_relation = avg_family_percent - pivot_table_family.loc['Не женат / не замужем', 'percent_in_time'] # вычислим разницу между между людьми состоявшими в браке и не никогда не состоявших в браке.

print(f"Люди никода не состоящие в браке, возвращают кредиты в срок реже на {round(family_status_relation, 2)}%, чем в среднем , люди, состоящие или бывшие в браке") # выведем результат на экран

Люди никода не состоящие в браке, возвращают кредиты в срок реже на 2.12%, чем в среднем , люди, состоящие или бывшие в браке


**Вывод:** Исходя из представленных расчетов, зависимость между семейным положением и возвратом кредита в срок существует, и в нашем случае, люди никода не состоящие в браке, возвращают кредиты в срок реже на **2.12%**, чем в среднем , люди, состоящие или бывшие в браке. 

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

 С выводом согласен, тебе верно удалось определить главный признак, по которому можно разделить все группы на 2 категории. 👍
    
 Если захочешь округлить значения столбца до нужного количества знаков после запятой, можно воспользоваться функций .round().
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [54]:
pivot_table_income = data.pivot_table(index='total_income_category', columns='debt', values='total_income', aggfunc='count') # составим сводную таблицу

pivot_table_income.columns = ["in_time", "not_in_time"] # переименуем колонки для удобства
pivot_table_income['median_salary'] = data.groupby('total_income_category')['total_income'].median() # добавим столбец, сгруппированый по группе дохода, с рассчитаным медианным уровнем дохода

total_loans_income = pivot_table_income['in_time'] + pivot_table_income['not_in_time'] # сохраним общее число займов
pivot_table_income.insert(0, 'total_loans', total_loans_income) # вcтавим общее кол-во займов вначало таблицы

pivot_table_income # выведем результат на экран

total_loans  in_time  not_in_time  median_salary
total_income_category                                                  
A                               25       23            2        1223042
B                             5014     4660          354         257558
C                            15921    14568         1353         130355
D                              349      328           21          43929
E                               22       20            2          26431

#### Мы вычислили медианное значение дохода для каждой группы дохода. Теперь мы можем оценить зависимость от уровня дохода и возвратом кредита в срок. Для этого расситаем процент возвращенных кредитов в срок и сделаем соответствующие выводы.

In [77]:
pivot_table_income['percent_in_time'] = (pivot_table_income['in_time'] / pivot_table_income['total_loans']) * 100 # процент возвращенных во время кредитов
pivot_table_income['percent_not_in_time'] = (pivot_table_income['not_in_time'] / pivot_table_income['total_loans']) * 100 # процент просроченых кредитов

total_loans_overall = pivot_table_income['total_loans'].sum() # кол-во всех выданных кредитов
pivot_table_income['loan_share_percent'] = round((pivot_table_income['total_loans'] / total_loans_overall) * 100, 2) # доля кредитов для каждой группы, в общем объеме выданных кредитов

pivot_table_income # вывод таблицы на экран

total_loans  in_time  not_in_time  median_salary  \
total_income_category                                                     
A                               25       23            2        1223042   
B                             5014     4660          354         257558   
C                            15921    14568         1353         130355   
D                              349      328           21          43929   
E                               22       20            2          26431   

                       percent_in_time  percent_not_in_time  \
total_income_category                                         
A                            92.000000             8.000000   
B                            92.939769             7.060231   
C                            91.501790             8.498210   
D                            93.982808             6.017192   
E                            90.909091             9.090909   

                       loan_share_percent  
total_income_category                      
A                                    0.12  
B                                   23.51  
C                                   74.64  
D                                    1.64  
E                                    0.10

In [78]:
summary_table = pivot_table_income[['median_salary', 'total_loans', 'in_time', 'not_in_time', 'percent_not_in_time', 'loan_share_percent']]

summary_table = summary_table.sort_values(by='percent_not_in_time', ascending=True)

summary_table

median_salary  total_loans  in_time  not_in_time  \
total_income_category                                                     
D                              43929          349      328           21   
B                             257558         5014     4660          354   
A                            1223042           25       23            2   
C                             130355        15921    14568         1353   
E                              26431           22       20            2   

                       percent_not_in_time  loan_share_percent  
total_income_category                                           
D                                 6.017192                1.64  
B                                 7.060231               23.51  
A                                 8.000000                0.12  
C                                 8.498210               74.64  
E                                 9.090909                0.10

**Вывод:** Исходя из расчетов, мы видим что заёмщики, относящиеся к группе **"D"** имеют самый низкий процент просроченых кредитов, равный **6%**. Заёмщики, относящиеся к самой многочисленной группе **"С"** имеют самое большое количество просрочек, и составляют основную массу заёмщиков, чья доля среди всех занимает почти 75%. Заёмщики в категории **"Е"** составляют одну из самых малочисленных групп, которую мы не будем брать в расчёт. Группа "B", вторая по объему выданных кредитов, имеет также один из самых низких процентов просроченых кредитов. В это время группа "А", с одним из самых высоких доходов, имеет ровно 8 процетов просроченых платежей. Несмотря на то что данная группа малочислена, займы, которые им могут выдаваться, могут быть гораздо больше по объему, нежели чем у других заёмщиков и иметь значительный оборот средств. (Для определения влияния на скоринговую систему следует получить больше данных о средсвах, которые были выданы группе "А") 

*Есть ли зависимость между уровнем дохода и возвратом кредита в срок?* **Да, есть**

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Верно. Правильно, что обращаешь внимание на объёмы категорий клиентов. 
    
Факультативно, можно попробовать решить задание альтернативным способом, разделив выборку на равные по объёму части, в данном случае, по столбцу сумм дохода. Для этого можно воспользоваться функцией pd.qcut().
    
Здесь можно об этом почитать:
    
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html

Метод qcut() делит числовой столбец на равные по объёму части на основании квантилей. Квантили — это статистические показатели, которые разделяют набор данных на равные части. Квантили включают квартили (25%, 50% и 75%), децили (каждые 10%) и процентили (каждые 1%). Приведу пример разделения на 3 равных отрезка, находящихся на отметках 33,3% и 66,7%.
        
    pd.qcut(data['column'], q=3, labels=['one', 'two', 'three'])
        
Можно не указывать названия отрезков с помощью labels. Тогда названия отрезкам присвоятся в численном виде.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [94]:
pivot_purpose = data.pivot_table(index='purpose_category', columns='debt', values='total_income', aggfunc='count' ) # Создём сводную таблицу
pivot_purpose.columns = ['in_time', 'not_in_time'] # переименовываем колонки для удобства.

pivot_purpose.sort_values(by='in_time', ascending=False) # сортируем значения по убыванию
total_loans_purpose = pivot_purpose['in_time'] + pivot_purpose['not_in_time'] # сохраняем значение всех выданных займов

pivot_purpose.insert(0, 'total_loans', total_loans_purpose) # добавляем колонку с обзим кол-вом займов вначало таблицы
pivot_purpose # выводим таблицу на экран

total_loans  in_time  not_in_time
purpose_category                                           
операции с автомобилем           4279     3879          400
операции с недвижимостью        10751     9971          780
получение образования            3988     3619          369
проведение свадьбы               2313     2130          183

In [99]:
pivot_purpose['percent_in_time'] = (pivot_purpose['in_time'] / pivot_purpose['total_loans']) * 100 # считаем процент кредитов возвращенных в срок
pivot_purpose['percent_not_in_time'] = (pivot_purpose['not_in_time'] / pivot_purpose['total_loans']) * 100 # считаем процент просроченых кредитов

pivot_purpose.sort_values(by='percent_in_time', ascending=False)

total_loans  in_time  not_in_time  percent_in_time  \
purpose_category                                                               
операции с недвижимостью        10751     9971          780        92.744861   
проведение свадьбы               2313     2130          183        92.088197   
получение образования            3988     3619          369        90.747242   
операции с автомобилем           4279     3879          400        90.652022   

                          percent_not_in_time  
purpose_category                               
операции с недвижимостью             7.255139  
проведение свадьбы                   7.911803  
получение образования                9.252758  
операции с автомобилем               9.347978

**Вывод:** Чаще всего, заёмщики отдают кредиты в срок **в операциях с недвижимостью** или на проведение **свадебных мероприятия**. Можно предположить что в первом случае, недвижимость очень важна для покрытия базовых потребностей и часто является необходимостью, что мотивирует заёмщиков платить в срок. Во втором случае, можно предположить что суммы займов не сильно существенны, и вероятно, берутся в дополнение к накоплениям. 

В случае же автокредитования, доля просрочек приближается к 10%. Это может быть обусловлено множеством факторов: от поломки автомобиля, на который был сделан займ, до угона, что очень мильно демотивирует заёмщика, отдавать кредит в срок, за то чем он не может пользоваться.

В категории "получение образования" также большой процент просрочек. Это также может быть связано со множеством факторов: от сложности программы обучения, до нежелания продолжать учёбу. Эти факторы также могут демотивировать заёмщиков. 

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Вывод верный. Здорово, что стараешься описывать причины, которые могли привести к полученной картине. Для аналитика важно видеть смысл, который стоит за данными.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* В нашем случае были пропущены значения в колонках "total_income"(совокупный доход за месяц) и "days_employed"(трудовой стаж). Вероятно, что пропуски в этих данных могли образоваться по следующим причинам:

1. Некоторые клиенты могла не предоставить данную информацию, или подтвердить её документально.
2. Неправильное заполнение анкет. Клиенты могли допустить ошибки при заполнении заявки на кредит.
3. Технические ошибки, которые могли привести к неправильной выгрузке данных. 


<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Абсолютно верно. Основные причины пропусков - это технические проблемы и человеческий фактор.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* В нашем случае, мы использовали медианное значение в вычислении зарплаты. В таком случае мы невелируем слишком высокие и слишком низкие значения дохода, которые могут помешать дальнейшим расчетам.

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Верно, медиана была выбрана именно по причине устойчивости к выбросам
</div>

### Шаг 4: общий вывод.

Мы выявили необходимые зависимости между запрашиваемыми данными. В итоге, мы видим примерный "портрет" заёмщиков. Внесение корректировок, основываясь на полученных данных, поможет в более точной настройке скоринговой системы, что в конечном счете **поможет снизить риски** при выдаче новых займов и кредитов, и, соответственно, **увеличить доходность.** Подробнее далее

### 1. Предобработка данных

Во время первоначального исследования данных, мы обнаружили пропуски в колонках "total_income"(совокупный доход за месяц) и "days_employed" (трудовой стаж). Вероятно, что пропуски в этих данных могли образоваться по следующим причинам:

  А) Некоторые клиенты могла не предоставить данную информацию, или подтвердить её документально.
  Б) Неправильное заполнение анкет. Клиенты могли допустить ошибки при заполнении заявки на кредит.
  В) Технические ошибки, которые могли привести к неправильной выгрузке данных.

Также, нельзя исключить что все три причины влияют на появление пропусков. Рекомендуется провести дополнительные проверки. 

В исследовании зависимости между количеством детей и возвратом кредита в строк были обнаружены аномальные значения, которые были удалены. Возможно, была допущена ошибка при заполнении анкеты или опечатка, допущенная сотрудником банка.

Зарплатные данные были успешно заменены медианными значениями, что позволило избежать резких отклонений в расчетах. 

Также виднеются разночтения во внесении информации о целях кредита. Судя по всему, в CRM систему данные в это поля вводятся вручную. Рекомендуем автоматизировать поле в которое будет вноситься данная информация с помощью выпадающего списка. Нами было определены следующие цели кредитования:

1. Операции с автомобилем
2. Операции с недвижимостью
3. Проведение свадьбы
4. получение образования 

### 2. Категоризация данных 

Мы произвели категоризацию заёмщиков по нескольким критериям: 

1. Количество детей 
2. Уровень дохода.
3. Цели кредитования. 
4. Семейное положение

Категоризация позволила нам находить взаимосвязь между разными факторами. 

### 3. Цели проекта

Нам необходимо было провести анализ данных и ответить на следующие вопросы:

1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
		
**Да, она присутствует**. Также было определено что в среднем, заёмщики с детьми выплачивают кредиты в срок на **1.6%** **реже** чем без детей. 

2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
		
Эта зависимость **также присутствует**. Заемщики, никода не состоящие в браке, возвращают кредиты в срок реже на **2.12%**, чем в среднем , заёмщики, состоящие или бывшие в браке. 

3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
		
**Да, зависимость есть**. Приведу вывод из исследования:
		
Исходя из расчетов, мы видим что заёмщики, относящиеся к группе **"D"** имеют самый низкий процент просроченных кредитов, равный **6%**. Заёмщики, относящиеся к самой многочисленной группе **"С"** имеют самое большое количество просрочек, и составляют основную массу заёмщиков, чья доля среди всех занимает почти 75%. Заёмщики в категории **"Е"** составляют одну из самых малочисленных групп, которую мы не будем брать в расчёт. Группа "B", вторая по объему выданных кредитов, имеет также один из самых низких процентов просроченых кредитов. В это время группа "А", с одним из самых высоких доходов, имеет ровно 8 процетов просроченых платежей. Несмотря на то что данная группа малочислена, займы, которые им могут выдаваться, могут быть гораздо больше по объему, нежели чем у других заёмщиков и иметь значительный оборот средств. (Для определения влияния на скоринговую систему следует получить больше данных о средсвах, которые были выданы группе "А")

4.  Как разные цели кредита влияют на его возврат в срок?

	Чаще всего, в срок ,заёмщики возвращают займы, связанные с **недвижимостью**( 7.25% просрочек ) и свадебным **мероприятиями**( 7.91%  просрочек ). **Получение образования** (9.25% просрочек) и **операции с автомобилем** (9.34% просрочек), возвращаются реже. 

	 Можно предположить что в первом случае, недвижимость очень важна для покрытия базовых потребностей и часто является необходимостью, что мотивирует заёмщиков платить в срок. Во втором случае, можно предположить что суммы займов не сильно существенны, и вероятно, берутся в дополнение к накоплениям.

	В случае же автокредитования, доля просрочек приближается к 10%. Это может быть обусловлено множеством факторов: от поломки автомобиля, на который был сделан займ, до угона, что очень сильно демотивирует заёмщика, отдавать кредит в срок, за то чем он не может пользоваться.

	В категории "получение образования" также большой процент просрочек. Это также может быть связано со множеством факторов: от сложности программы обучения, до нежелания продолжать учёбу. Эти факторы также могут демотивировать заёмщиков.


### 4. Рекомендации

Исходя из проведенного анализа, можно представить следующие рекомендации:

1. **Улучшение сбора данных**:

	- Обеспечить более тщательную проверку анкет заемщиков, чтобы минимизировать пропуски и ошибки при вводе данных.
	- Автоматизировать процессы ввода информации о целях кредитования с помощью выпадающих списков и других средств, исключающих ручной ввод.

2. **Мониторинг и оценка рисков**:

	- Создать систему регулярного мониторинга и анализа ключевых показателей эффективности (KPI) по различным категориям заемщиков и типам кредитов.
	

3. **Адаптация кредитных предложений**:

	- Разработать кредитные предложения в зависимости от целей кредитования. Например, усилить контроль и ужесточить условия для автокредитования и кредитов на образование, где уровень просрочек выше.
	- Предлагать более привлекательные условия для займов, связанных с недвижимостью и свадьбами, где уровень просрочек ниже.

Реализация этих рекомендаций поможет улучшить точность и надежность скоринговой системы, что в конечном итоге приведет к **снижению рисков**, повышению **доходности** и укреплению доверия клиентов.


### Предобработка данных

<div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

Теперь получился очень хороший структурированный вывод. Все этапы проекта рассмотрены, а категории клиентов учтены. И даны рекомендации для заказчика. Отличная работа!
    
</div>

<div class="alert alert-danger">
    <h3> Комментарий ревьюера ❌<a class="tocSkip"> </h3>
    
Чтобы твою работу заметили и ничего не упустили, важно описывать всё в деталях.
        
Стоит учитывать, что твой потенциальный руководитель, в первую очередь, обратит внимание на общий вывод. Поэтому важно отразить в нём все основные результаты из всех шагов проекта в числах. И важно описать проблемы, найденные на этапе предобработки, и пути их решения
        
Можно опираться на такую структуру финального вывода во всех проектах:
* предобработка данных с описанием найденных проблем и путей их решения
* ответы на цели проекта
* рекомендации для заказчика

Чтобы лучше понимать, как оформлять комментарии в markdown-ячейках, советую изучить [краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/).

</div>